### PR TITLE
feat: 2D slab correction for Ewald summation

### PR DESCRIPTION
**Pull request by atulcthakur**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Adds 2D slab correction support for Ewald summation, including triclinic slab cells, analytical forces/charge gradients/virial, and hybrid-force compatibility. 
Adds torch bindings for slab kernels. 

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- Added standalone `apply_slab_correction(...)` and integrated `ewald_summation(..., slab_correction=True, pbc=...)` support.
- Added Warp slab correction kernels for energy, forces, charge gradients, and virial.
- Added triclinic slab cell support using projected-normal geometry.
- Added hybrid-force compatibility for slab corrections.
- Added CPU/GPU tests for slab kernels, Torch bindings, triclinic cells, autograd validation, and TorchPME cross-validation.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [ ] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/78/commits/afa5c0d66b970f3e6933f6f1b4727fb04a664116

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/78/commits/afa5c0d66b970f3e6933f6f1b4727fb04a664116

**Comment by atulcthakur** [comment]
/ok to test [6fe7a4a](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/78/commits/6fe7a4ae66a66553c96fce753cb7b37d9b7dce8a)

File: nvalchemiops/interactions/electrostatics/slab_kernels.py [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+            wp.array2d(dtype=wp.float64),  # mz (B, 3)
+            wp.array2d(dtype=wp.float64),  # mz2 (B, 3)
+            wp.array(dtype=wp.float64),  # qtotal
+            wp.array(dtype=wp.float64),  # energy_in
+            wp.array(dtype=wp.float64),  # energy_out
+            wp.array(dtype=v),  # forces
+            wp.array(dtype=wp.float64),  # charge_grads
+            wp.array(dtype=m),  # virial
+        ],
+    )
```
[the remark this answers is not in the document]
Comment (dallasfoster) [reply nvalchemiops/interactions/electrostatics/slab_kernels.py]: **Issue:** The kernel signature accepts `forces`, `charge_grads`, and `virial` outputs, and the body unconditionally:
- Builds `f_slab` and atomic-adds at line 427.
- Computes `cg_slab` and atomic-adds at line 431.
- Builds the 3×3 `virial_mat` and atomic-adds at line 453.

There is no `compute_forces: wp.bool` / `compute_charge_gradients: wp.bool` / `compute_virial: wp.bool` runtime flag, unlike every other kernel in the module. For comparison:
- `nvalchemiops/interactions/electrostatics/ewald_kernels.py:4459` — `_ewald_real_space_energy_forces_kernel` takes `wp.bool` for `compute_virial` and only emits the virial write when set.
- `ewald_kernels.py:4513` — same pattern for the charge-grad kernel.
- The existing module also splits energy-only / +forces / +charge_grads into separate overloads (e.g. `_ewald_real_space_energy_kernel` vs `_ewald_real_space_energy_forces_kernel` vs `_ewald_real_space_energy_forces_charge_grad_kernel`).

**Consequence:**
1. The energy-only path (`compute_forces=False, compute_charge_gradients=False, compute_virial=False`) still pays the per-atom force/CG/virial FLOPs and 3 atomic_adds. Cost is small relative to Ewald but real, and the unnecessary atomic-add contention is the slowest part.
2. `_slab_correction_op` in `ewald.py:3232-3263` always allocates `slab_forces`, `slab_charge_grads`, `slab_virial` even when the caller asked for energy only. These are wasted `torch.zeros` allocations.
3. Stylistic divergence from the rest of the module — same operation, three different conventions.

**Suggested fix (pick one):**
- ** (matches the rest of Ewald):** split into `_slab_correction_e_kernel`, `_slab_correction_ef_kernel`, `_slab_correction_efcv_kernel` overloads, and have `_slab_correction_op` dispatch + allocate only what's needed. This is what every other section of `ewald_kernels.py` does.

File: test/interactions/electrostatics/bindings/torch/test_slab.py
Comment (dallasfoster) [inline test/interactions/electrostatics/bindings/torch/test_slab.py:1]: **Suggested verification test:** add to `test_slab.py::TestSlabHybridForces`:
```python
def test_hybrid_slab_returned_charge_grads_match_analytical(self, device):
    """Returned charge_grads should equal sum(rs, rec, slab) — not 2x slab."""
    # standard mode
    e_std, cg_std = ewald_summation(..., compute_charge_gradients=True, pbc=pbc, slab_correction=True)
    # hybrid mode, same args
    e_hyb, cg_hyb = ewald_summation(..., compute_charge_gradients=True, hybrid_forces=True, pbc=pbc, slab_correction=True)
    torch.testing.assert_close(cg_hyb, cg_std, rtol=1e-10, atol=1e-12)

**Review by dallasfoster — COMMENTED** [review]
No example file, no benchmark entry, no docs update
**Files:**
- `examples/electrostatics/` — currently has `01_coulomb_example.py` through `04_jax_pme_example.py`, no slab example.
- `benchmarks/interactions/electrostatics/benchmark_electrostatics.py` — no slab variant.
- `docs/` — no mention of `apply_slab_correction` or `slab_correction=` flag.

**Why this matters:** the new feature has no walkthrough for end users. Compare with `examples/electrostatics/02_ewald_summation_example.py:155-171` which is the canonical "how to use Ewald" demo — slab correction needs an analogous file. Without it, the first downstream consumer will have to read the test file to figure out the API.

File: nvalchemiops/interactions/electrostatics/slab_kernels.py [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+
+    if (not p0) and p1 and p2:
+        axis_idx = wp.int32(0)
+        is_slab = True
+    elif p0 and (not p1) and p2:
+        axis_idx = wp.int32(1)
+        is_slab = True
+    elif p0 and p1 and (not p2):
+        axis_idx = wp.int32(2)
+        is_slab = True
```
Comment (dallasfoster) [inline nvalchemiops/interactions/electrostatics/slab_kernels.py:214]: Per-system geometry recomputed per atom
**File:** `nvalchemiops/interactions/electrostatics/slab_kernels.py:201-247` (reduce kernel) and `347-397` (correction kernel)

Both kernels compute `axis_idx`, pick the two periodic vectors, and compute `normal` per atom (i.e. N times per system, redundantly). For axis-aligned cells this is cheap, but for triclinic + large N it's measurable bandwidth (each thread reads the same `cell[system_id]` mat33).

**Suggested optimization (low priority):** add a small `_slab_precompute_geometry_kernel` (`dim=B`) that writes `(n_x, n_y, n_z, axis_idx, vol, L²)` per system into a (B, 6) workspace; both other kernels read from it. Reduces per-atom work and lets you drop the `cell` argument from the correction kernel, simplifying the autograd tape. Skip if profiling shows this isn't a hot path.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py [hunk 3]
```diff
@@ -3125,6 +3131,328 @@ def _build_result(energies, forces=None, charge_grads=None, virial=None):
...
+            "directions and False for the non-periodic (slab) direction."
+        )
+    if pbc.dtype != torch.bool:
+        raise ValueError(f"`pbc` must be a bool tensor, got dtype={pbc.dtype}.")
+    if pbc.dim() == 1:
+        if pbc.shape[0] != 3:
+            raise ValueError(
+                f"`pbc` of shape (3,) expected, got shape {tuple(pbc.shape)}."
+            )
+        pbc = pbc.unsqueeze(0).expand(num_systems, 3).contiguous()
```
Comment (dallasfoster) [inline nvalchemiops/torch/interactions/electrostatics/ewald.py]: `_prepare_pbc_for_slab` materializes a B×3 tensor per call
**File:** `nvalchemiops/torch/interactions/electrostatics/ewald.py:3157-3162`

```python
pbc = pbc.unsqueeze(0).expand(num_systems, 3).contiguous()
```

`.contiguous()` after `.expand()` allocates a new B×3 tensor on every call. In a tight MD loop this is wasted work because pbc usually doesn't change step-to-step. Not blocking, but document that users can pass a `(B, 3)` pbc directly to skip this.

File: test/interactions/electrostatics/test_slab_kernels.py
Comment (dallasfoster) [inline test/interactions/electrostatics/test_slab_kernels.py:1]: Suggested additional tests:

torch.autograd.gradcheck on a small (N≤4) float64 system, all of {positions, charges, cell} requiring grad, all 4 outputs.
Empty system (N=0) — kernel launch with dim=0 should not crash.
Single-atom non-neutral — exercises the Q²L²/12 and Qz² terms without cancellation.
Mixed-pbc batch including a 3D-periodic system — confirms the 3D system contributes zero while slab systems compute correctly. Currently test_mixed_axis covers only slab+slab.

**Comment by atulcthakur** [comment]
> No example file, no benchmark entry, no docs update **Files:**
> 
> * `examples/electrostatics/` — currently has `01_coulomb_example.py` through `04_jax_pme_example.py`, no slab example.
> * `benchmarks/interactions/electrostatics/benchmark_electrostatics.py` — no slab variant.
> * `docs/` — no mention of `apply_slab_correction` or `slab_correction=` flag.
> 
> **Why this matters:** the new feature has no walkthrough for end users. Compare with `examples/electrostatics/02_ewald_summation_example.py:155-171` which is the canonical "how to use Ewald" demo — slab correction needs an analogous file. Without it, the first downstream consumer will have to read the test file to figure out the API.

Yes, these were intentionally left out to limit the scope and size of this PR. I can fold it in, no problem, but I was going to do a separate small PR at the end when both Ewald and PME slab interfaces get finalized and merged.

File: test/interactions/electrostatics/bindings/torch/test_slab.py
Comment (atulcthakur) [reply test/interactions/electrostatics/bindings/torch/test_slab.py:1]: > The kernel signature accepts forces, charge_grads, and virial outputs, and the body unconditionally:

I did think about this and made a decision to not implement 3 kernels as the rest of the Ewald module. IMO, doing so, exposes a lot of unnecessary surface area (and increases maintenance) for the relatively cheap slab ops. I was thinking to do branching within the kernel if that would work. IMO, the impact of this is likely negligible since slab correction itself only does O(N) work.

File: nvalchemiops/interactions/electrostatics/slab_kernels.py (hunk 2 above)
Comment (atulcthakur) [reply nvalchemiops/interactions/electrostatics/slab_kernels.py:214]: Noted.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py (hunk 3 above)
Comment (atulcthakur) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Noted.

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/78/commits/b2ad6600ecccbd879c67a78123617f534198d7c4

File: test/interactions/electrostatics/test_slab_kernels.py
Comment (atulcthakur) [reply test/interactions/electrostatics/test_slab_kernels.py:1]: Added.

File: nvalchemiops/torch/interactions/electrostatics/ewald.py (hunk 3 above)
Comment (atulcthakur) [reply nvalchemiops/torch/interactions/electrostatics/ewald.py]: Removed this. We do not silently expand now, as it can lead to unintended behavior. Users must pass in a contiguous (B, 3) pbc tensor explicitly.

File: nvalchemiops/interactions/electrostatics/slab_kernels.py (hunk 2 above)
Comment (atulcthakur) [reply nvalchemiops/interactions/electrostatics/slab_kernels.py:214]: Agreed, but this would need a third kernel launch. Deferring for now.

File: test/interactions/electrostatics/bindings/torch/test_slab.py
Comment (atulcthakur) [reply test/interactions/electrostatics/bindings/torch/test_slab.py:1]: Implemented Ewald style 3 kernels due to perf regressions on small (~500 atoms) systems.

Comment (atulcthakur) [reply test/interactions/electrostatics/bindings/torch/test_slab.py:1]: `test_hybrid_slab_returned_charge_grads_match_analytical` test added

File: docs/userguide/components/electrostatics.md [hunk 4]
```diff
@@ -484,6 +486,52 @@ k_{\text{cutoff}} = \sqrt{-2 \ln \varepsilon} / \eta
 Refer to the [Parameter Estimation](parameter-estimation) section for API usage.
 ```
 
+#### 2D Slab Correction (PyTorch)
+
+For interfacial systems with two periodic directions and one non-periodic direction,
+PyTorch Ewald can add the Yeh-Berkowitz / Ballenegger slab correction. Pass
```
Comment (laserkelvin) [inline docs/userguide/components/electrostatics.md:492]: Do you want to provide a reference link for this?

**Review by laserkelvin — APPROVED** [review]
Kernel and interface generally looks good to me

File: docs/userguide/components/electrostatics.md (hunk 4 above)
Comment (atulcthakur) [reply docs/userguide/components/electrostatics.md:492]: Sure. Will add a citation in the next PR.
